File: hark-asr/src/utils/markdownPlain.ts
/**
 * Markdown → 朗读文本：不念格式符号，按结构插入停顿。
 *
 * 停顿约定（两端 TTS 都认标点）：
 * - 标题后：较长停顿（……）
 * - 段落 / 分隔线：句号停顿（。）
 * - 列表项之间：分号停顿（；）
 * - 行内强调等：只保留文字，不加额外停顿
 */

const LONG = "……";
const MED = "。";
const SHORT = "；";

function stripInline(s: string): string {
  let t = s;
  t = t.replace(/!\[([^\]]*)\]\([^)]+\)/g, "$1");
  t = t.replace(/\[([^\]]+)\]\([^)]+\)/g, "$1");
  t = t.replace(/\[([^\]]+)\]\[[^\]]*\]/g, "$1");
  t = t.replace(/\[\^[^\]]+\]/g, "");
  t = t.replace(/(\*\*|__)(.*?)\1/g, "$2");
  t = t.replace(/(\*|_)(.*?)\1/g, "$2");
  t = t.replace(/~~(.*?)~~/g, "$1");
  t = t.replace(/`([^`]+)`/g, "$1");
  t = t.replace(/<\/?[a-zA-Z][^>]*>/g, "");
  return t.replace(/[ \t]{2,}/g, " ").trim();
}

function endsWithPause(s: string): boolean {
  return /[。！？!?…；;，,：:]$/.test(s.trim());
}

function joinPause(parts: string[]): string {
  return parts
    .map((p) => p.trim())
    .filter(Boolean)
    .join("")
    .replace(/。{2,}/g, "。")
    .replace(/；{2,}/g, "；")
    .replace(/，{2,}/g, "，")
    .replace(/(……)+/g, "……")
    .replace(/\s+/g, " ")
    .trim();
}

/** 将 Markdown（或纯文本）转为适合朗读的正文，按格式给停顿。 */
export function markdownToSpeech(input: string): string {
  let raw = input.replace(/\r\n/g, "\n");
  raw = raw.replace(/^---\n[\s\S]*?\n---\n?/, "");
  raw = raw.replace(/<!--[\s\S]*?-->/g, "");

  const parts: string[] = [];
  const lines = raw.split("\n");
  let i = 0;
  let inFence = false;

  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();

    // fenced code：跳过内容，给一次中等停顿
    if (/^```|^~~~/.test(trimmed)) {
      if (!inFence) {
        inFence = true;
      } else {
        inFence = false;
        parts.push(MED);
      }
      i += 1;
      continue;
    }
    if (inFence) {
      i += 1;
      continue;
    }

    // blank line → 段落停顿
    if (!trimmed) {
      if (parts.length && !endsWithPause(parts[parts.length - 1] ?? "")) {
        parts.push(MED);
      }
      i += 1;
      continue;
    }

    // horizontal rule
    if (/^(-{3,}|\*{3,}|_{3,})$/.test(trimmed)) {
      parts.push(LONG);
      i += 1;
      continue;
    }

    // heading
    const heading = trimmed.match(/^(#{1,6})\s+(.*)$/);
    if (heading) {
      const level = heading[1].length;
      const text = stripInline(heading[2]);
      if (text) {
        parts.push(text);
        parts.push(level <= 2 ? LONG : MED);
      }
      i += 1;
      continue;
    }

    // blockquote
    if (/^>\s?/.test(trimmed)) {
      const text = stripInline(trimmed.replace(/^(>\s?)+/, ""));
      if (text) {
        parts.push(text);
        if (!endsWithPause(text)) parts.push(MED);
      }
      i += 1;
      continue;
    }

    // table separator — skip
    if (/^\|?[\t ]*:?-{3,}:?[\t ]*(\|[\t ]*:?-{3,}:?[\t ]*)+\|?\s*$/.test(trimmed)) {
      i += 1;
      continue;
    }

    // table row
    if (/^\|.+\|/.test(trimmed)) {
      const cells = trimmed
        .replace(/^\|/, "")
        .replace(/\|$/, "")
        .split("|")
        .map((c) => stripInline(c))
        .filter(Boolean);
      if (cells.length) {
        parts.push(cells.join("，"));
        parts.push(SHORT);
      }
      i += 1;
      continue;
    }

    // list / task
    const list = trimmed.match(/^([-*+]|\d+\.)\s+(?:\[[ xX]\]\s+)?(.*)$/);
    if (list) {
      const text = stripInline(list[2]);
      if (text) {
        parts.push(text);
        parts.push(endsWithPause(text) ? "" : SHORT);
      }
      i += 1;
      continue;
    }

    // reference / footnote definition — skip
    if (/^\[([^\]]+)\]:\s+\S+/.test(trimmed) || /^\[\^[^\]]+\]:/.test(trimmed)) {
      i += 1;
      continue;
    }

    // normal paragraph line
    const text = stripInline(trimmed);
    if (text) {
      parts.push(text);
      // 若下一行是空行或结构块，稍后空白处理会加句号；同行续写加空格
      const next = lines[i + 1]?.trim() ?? "";
      if (!next) {
        if (!endsWithPause(text)) parts.push(MED);
      } else if (/^(#{1,6}\s|([-*+]|\d+\.)\s|>|```|~~~|(-{3,}|\*{3,}|_{3,})$)/.test(next)) {
        if (!endsWithPause(text)) parts.push(MED);
      } else {
        parts.push(" ");
      }
    }
    i += 1;
  }

  return joinPause(parts);
}

File: hark-swift/Hark/Utils/MarkdownPlain.swift
import Foundation

/// Markdown → 朗读文本：不念格式符号，按结构插入停顿（移植 markdownPlain.ts）。
/// 标题后较长停顿（……），段落/分隔线句号（。），列表项间分号（；）。
enum MarkdownPlain {
    private static let long = "……"
    private static let med = "。"
    private static let short = "；"

    private static let imageLink = /!\[([^\]]*)\]\([^)]+\)/
    private static let inlineLink = /\[([^\]]+)\]\([^)]+\)/
    private static let refLink = /\[([^\]]+)\]\[[^\]]*\]/
    private static let footnoteRef = /\[\^[^\]]+\]/
    private static let bold = /(\*\*|__)(.*?)\1/
    private static let italic = /(\*|_)(.*?)\1/
    private static let strike = /~~(.*?)~~/
    private static let code = /`([^`]+)`/
    private static let htmlTag = /<\/?[a-zA-Z][^>]*>/
    private static let multiSpace = /[ \t]{2,}/

    private static let pauseEnd = /[。！？!?…；;，,：:]$/
    private static let heading = /^(#{1,6})\s+(.*)$/
    private static let hr = /^(-{3,}|\*{3,}|_{3,})$/
    private static let blockquote = /^>\s?/
    private static let tableSeparator = /^\|?[\t ]*:?-{3,}:?[\t ]*(\|[\t ]*:?-{3,}:?[\t ]*)+\|?\s*$/
    private static let tableRow = /^\|.+\|/
    private static let listItem = /^([-*+]|\d+\.)\s+(?:\[[xX ]\]\s+)?(.*)$/
    private static let refDef = /^\[([^\]]+)\]:\s+\S+/
    private static let footnoteDef = /^\[\^[^\]]+\]:/
    private static let fence = /^(```|~~~)/
    private static let structural = /^(#{1,6}\s|([-*+]|\d+\.)\s|>|```|~~~|(-{3,}|\*{3,}|_{3,})$)/

    private static func stripInline(_ s: String) -> String {
        var t = s
        t = t.replacing(imageLink) { String($0.1) }
        t = t.replacing(inlineLink) { String($0.1) }
        t = t.replacing(refLink) { String($0.1) }
        t = t.replacing(footnoteRef, with: "")
        t = t.replacing(bold) { String($0.2) }
        t = t.replacing(italic) { String($0.2) }
        t = t.replacing(strike) { String($0.1) }
        t = t.replacing(code) { String($0.1) }
        t = t.replacing(htmlTag, with: "")
        return t.replacing(multiSpace, with: " ").trimmingCharacters(in: .whitespaces)
    }

    private static func endsWithPause(_ s: String) -> Bool {
        s.firstMatch(of: pauseEnd) != nil
    }

    private static func joinPause(_ parts: [String]) -> String {
        parts
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
            .joined()
            .replacing(/。{2,}/, with: "。")
            .replacing(/；{2,}/, with: "；")
            .replacing(/，{2,}/, with: "，")
            .replacing(/(……)+/, with: "……")
            .replacing(/\s+/, with: " ")
            .trimmingCharacters(in: .whitespaces)
    }

    static func toSpeech(_ input: String) -> String {
        var raw = input.replacingOccurrences(of: "\r\n", with: "\n")
        // YAML frontmatter
        if let range = raw.range(of: "^---\n[\\s\\S]*?\\n---\\n?", options: .regularExpression) {
            raw.replaceSubrange(range, with: "")
        }
        raw = raw.replacing(/<!--[\s\S]*?-->/, with: "")

        var parts: [String] = []
        let lines = raw.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        var i = 0
        var inFence = false

        while i < lines.count {
            let line = lines[i]
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            if trimmed.firstMatch(of: fence) != nil {
                if !inFence {
                    inFence = true
                } else {
                    inFence = false
                    parts.append(med)
                }
                i += 1
                continue
            }
            if inFence {
                i += 1
                continue
            }

            if trimmed.isEmpty {
                if let last = parts.last, !endsWithPause(last) {
                    parts.append(med)
                }
                i += 1
                continue
            }

            if trimmed.firstMatch(of: hr) != nil {
                parts.append(long)
                i += 1
                continue
            }

            if let h = trimmed.firstMatch(of: heading) {
                let level = h.1.count
                let text = stripInline(String(h.2))
                if !text.isEmpty {
                    parts.append(text)
                    parts.append(level <= 2 ? long : med)
                }
                i += 1
                continue
            }

            if trimmed.firstMatch(of: blockquote) != nil {
                let text = stripInline(trimmed.replacing(blockquoteStart, with: ""))
                if !text.isEmpty {
                    parts.append(text)
                    if !endsWithPause(text) { parts.append(med) }
                }
                i += 1
                continue
            }

            if trimmed.firstMatch(of: tableSeparator) != nil {
                i += 1
                continue
            }

            if trimmed.firstMatch(of: tableRow) != nil {
                var t = trimmed
                if t.hasPrefix("|") { t.removeFirst() }
                if t.hasSuffix("|") { t.removeLast() }
                let cells = t.split(separator: "|", omittingEmptySubsequences: true)
                    .map { stripInline(String($0)) }
                    .filter { !$0.isEmpty }
                if !cells.isEmpty {
                    parts.append(cells.joined(separator: "，"))
                    parts.append(short)
                }
                i += 1
                continue
            }

            if let list = trimmed.firstMatch(of: listItem) {
                let text = stripInline(String(list.2))
                if !text.isEmpty {
                    parts.append(text)
                    if !endsWithPause(text) { parts.append(short) }
                }
                i += 1
                continue
            }

            if trimmed.firstMatch(of: refDef) != nil || trimmed.firstMatch(of: footnoteDef) != nil {
                i += 1
                continue
            }

            let text = stripInline(trimmed)
            if !text.isEmpty {
                parts.append(text)
                let next = i + 1 < lines.count ? lines[i + 1].trimmingCharacters(in: .whitespaces) : ""
                if next.isEmpty {
                    if !endsWithPause(text) { parts.append(med) }
                } else if next.firstMatch(of: structural) != nil {
                    if !endsWithPause(text) { parts.append(med) }
                } else {
                    parts.append(" ")
                }
            }
            i += 1
        }

        return joinPause(parts)
    }
}

private let blockquoteStart = /^(>\s?)+/
